Issue676

Zeglow opened the pull request:
# Homework 2 Pull Request
## Checklist before creating
- [x] There are no Checkstyle issues.
- [x] There are no SpotBugs issues.
- [x] There are no javadoc errors.
- [x] All tests in the file pass.
## Questions
### Will you get full credit on this assignment?
Yes, I expect to receive full credit. I successfully converted 15 test methods with multiple assertions to use @ParameterizedTest, following JUnit 5 best practices. 
### How long did you spend on the assignment?
Approximately 6 hours total, including:
- Initial setup: It took so long because there was something wrong with jbang when I run build jabref, I tried multiple methods but didn't fix it. Then it recovered itself... Maybe it was caused by network issues. (2 hours)
- Converting the 15 test methods (2.5 hours)  
- Testing (0.5 hour)
- Commit&PR (0.5 hour)
### What problems did you encounter, and how did you deal with them?
JBang had problems during the initial `build jabref` step. After trying multiple solutions, it resolved itself, likely due to network connectivity issues. This delayed the initial setup by about 2 hours.

Initially I ran tests in the wrong module (jabgui instead of jablib). Solved by identifying the correct file location.
### What help did you give or receive on this assignment (including answering these question)? What was your AI usage?
I collaborated with a classmate on task division - she handled the first 11 methods and the other class in AB while I converted the remaining 15. We coordinated to avoid duplicate work.
I used AI assistance (Claude) for:
- Learning proper @ParameterizedTest syntax and best practices
- Code review suggestions
- Git commands suggestions
- Debugging gradle test execution issues
---
## Summary of Changes
Methods converted: nAuthors1, applyModifiers, checkLegalKeyNoUnwantedCharacters, checkLegalKeyUnwantedCharacters, keywordsNKeywordsSeparatedBySpace, keywordNKeywordsSeparatedBySpace, title, camel, shortTitle, veryShortTitle, lastPage, pagePrefix, firstPage, nAuthors3, oneAuthorPlusIni

Zeglow committed:
Convert nAuthors3() to @ParameterizedTest

- Converted nAuthors3() test method to use @ParameterizedTest with @MethodSource

Addresses #676

Zeglow committed:
Convert firstPage() to @ParameterizedTest - Converted firstPage() test method to use @ParameterizedTest with @MethodSource Addresses #676

Zeglow committed:
Convert pagePrefix() to @ParameterizedTest

Zeglow committed:
Convert lastPage() to @ParameterizedTest

Zeglow committed:
convert veryShortTitle to @ParameterizedTest

Zeglow committed:
convert shortTitle to @ParameterizedTest

Zeglow committed:
Convert title() to @ParameterizedTest

Zeglow committed:
Convert keywordNKeywordsSeparatedBySpace() to @ParameterizedTest

Zeglow committed:
Convert keywordsNKeywordsSeparatedBySpace() to @ParameterizedTest

Zeglow committed:
Convert checkLegalKeyUnwantedCharacters() to @ParameterizedTest

Zeglow committed:
Convert applyModifiers() to @ParameterizedTest

Zeglow committed:
Convert checkLegalKeyNoUnwantedCharacters() to @ParameterizedTest

Zeglow committed:
Convert oneAuthorPlusIni() to @ParameterizedTest

Zeglow committed:
Convert nAuthors1() to @ParameterizedTest

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
```diff
@@ -831,18 +853,23 @@ void crossrefkeywordNKeywordsSeparatedBySpace() {
 
-    @Test
-    void keywordsNKeywordsSeparatedBySpace() {
+    // It seems this one is the same as keywordNKeywordsSeparatedBySpace
```
You can use it with both test methods. Include a comment saying the two test methods that use it.

Zeglow replied:
Thank you for the feedback! I apologize for the confusion. I realize now that these two test methods use different @MethodSource methods (keywordNData & keywordsNData). After reviewing the code more carefully, I think it might be better to keep them separate because: They test two distinct features. Also, keeping them separate makes it easier to identify which feature has issues if a test fails.

However, I'm not sure about this. I'm happy to refactor them to a single @MethodSource if that would be better for maintainability. Thanks!

espertusnu replied:
If they have exactly the same content (which isn't easy for me to see), create a single provider method and put a comment before it saying which two test methods use it.

If their content is different, keep them separate.

Zeglow replied:
Thank you. I'll keep them seperate.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
```diff
@@ -831,18 +853,23 @@ void crossrefkeywordNKeywordsSeparatedBySpace() {
 
-    @Test
-    void keywordsNKeywordsSeparatedBySpace() {
+    // It seems this one is the same as keywordNKeywordsSeparatedBySpace
```
I don't understand what you mean. Maybe we can discuss this.

espertusnu replied:
I don't remember discussing this with you. Could you refresh my memory or arrange a time to discuss it?

espertusnu commented:
What is the status of this PR?

Zeglow committed:
Merge main branch and resolve submodule conflicts

Zeglow commented:
> What is the status of this PR?

I just resolved the submodule conflicts (csl-locales and csl-styles). It seems that this PR has no conflicts with the base branch now.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
Why not parameterize these?

Zeglow replied:
Lucia and I both worked on the [CitationKeyGeneratorTest.java] to complete all the methods. I'll update my version with her changes as soon as possible.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
This can be combined with the previous method.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
Why not parameterize these?

espertusnu reviewed (changes requested):
Please look for additional opportunities to parameterize tests. Let me know if you have any questions.

espertusnu commented:
[like]  Spertus, Ellen reacted to your message:
________________________________
From: Avery ***@***.***>
Sent: Tuesday, October 7, 2025 5:16:11 PM
To: Zeglow/jabref ***@***.***>
Cc: Spertus, Ellen ***@***.***>; Comment ***@***.***>
Subject: Re: [Zeglow/jabref] Issue676 (PR #1)


@Zeglow commented on this pull request.

________________________________

In jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java<https://github.com/Zeglow/jabref/pull/1#discussion_r2411326067>:

>


Lucia and I both worked on the [CitationKeyGeneratorTest.java] to complete all the methods. I'll update my version with her changes as soon as possible.

—
Reply to this email directly, view it on GitHub<https://github.com/Zeglow/jabref/pull/1#discussion_r2411326067>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A44JAIZRQIF7EGHXW5IY4D33WPYNXAVCNFSM6AAAAACHD6ES5KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTGMJRGA4TANBUHE>.
You are receiving this because you commented.Message ID: ***@***.***>

espertusnu commented:
> Lucia and I both worked on the [CitationKeyGeneratorTest.java] to complete all the methods. I'll update my version with her changes as soon as possible.

I hadn't remembered that multiple people were working on that file. Just let me know what lines are your responsibility, and I'll review only those.

espertusnu commented:
It did seem unlike you not to do a thorough job!

Zeglow commented:
> > Lucia and I both worked on the [CitationKeyGeneratorTest.java] to complete all the methods. I'll update my version with her changes as soon as possible.
> 
> I hadn't remembered that multiple people were working on that file. Just let me know what lines are your responsibility, and I'll review only those.

Thank you! I worked on the methods starting from line 579 -> oneAuthorPlusIni.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
```diff
@@ -580,27 +580,40 @@ void lastAuthorForenameInitials() {
+        );
     }
 
     @Test
```
Why not combine this with the previous test?

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
```diff
@@ -879,12 +920,18 @@ void checkLegalNullInNullOut() {
         assertThrows(NullPointerException.class, () -> CitationKeyGenerator.cleanKey(null, DEFAULT_UNWANTED_CHARACTERS));
```
Is there any difference between these two lines?

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
Is there any reason you're not combining this with the following tests?

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
Is there any reason not to combine this with the next test?

Zeglow committed:
Address review comments: combine tests and remove duplicate assertions

Zeglow committed:
Address review comments: Combined nAuthors1EmptyReturnEmpty() into nAuthors1() parameterized test

Zeglow commented:
Hi Ellen,
I've addressed all the review comments. So sorry for the late response! I was unwell with an ear infection these days and wasn't feeling well enough to work...Thanks for your patience! 
P.S. Do I need to merge my changes with Lucia's modifications before submitting a PR to the main JabRef repository?
Thanks,
Avery

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
```diff
@@ -916,16 +958,18 @@ void crossrefShorttitleInitials() {
+                Arguments.of("Green Scheduling of `Whatever`", "GreenSchedulingofWhatever")
+        );
     }
 
```
This could be CSV.

espertusnu commented on jablib/src/test/java/org/jabref/logic/citationkeypattern/CitationKeyGeneratorTest.java:
This too.

espertusnu commented:
I'm sorry you were unwell and I hope you are all better.

While it would be easier for the jabref people if you merge your changes, one of you wouldn't be listed as a contributor.

Something you could if Lucia agreed would be to split on file boundaries, where you submit the 2 files you did alone, and she submits the file you both worked on, mentioning you in the PR. @Zeglow

Zeglow committed:
Refactor: Convert CitationKeyGeneratorTest methods to parameterized test

Zeglow commented:
> I'm sorry you were unwell and I hope you are all better.
> 
> While it would be easier for the jabref people if you merge your changes, one of you wouldn't be listed as a contributor.
> 
> Something you could if Lucia agreed would be to split on file boundaries, where you submit the 2 files you did alone, and she submits the file you both worked on, mentioning you in the PR. @Zeglow

Hi Ellen,

Thank you for your understanding! 

I discussed this with Lucia. Actually, I don't have any files that I worked on independently - we only collaborated on CitationKeyGeneratorTest.java together. So we've decided that I will submit the PR for CitationKeyGeneratorTest.java and mention her contributions in the PR description. And she will submit the file she worked on independently. This way we'll both be listed as contributors.

espertusnu reviewed (commented):
I see that I failed to share these comments. Don't bother correcting them in this PR, but do correct them.
